File: duplex/cpp/tests/seam.cpp
#include <nightseam/duplex/conn.hpp>

#include <chrono>
#include <future>
#include <iostream>
#include <stdexcept>

using namespace nightseam::duplex;
using namespace std::chrono_literals;

void require(bool condition, const char* message) {
    if (!condition) throw std::runtime_error(message);
}

template<class Error, class F> void refuses(F&& operation) {
    try { operation(); } catch (const Error&) { return; }
    throw std::runtime_error("operation did not refuse with the expected error");
}

void seam() {
    auto [a, b] = pipe();
    a->send({Kind::text, "first"});
    a->send({Kind::binary, std::string("\0\xff", 2)});
    b->send({Kind::text, "reverse"});
    auto first = b->receive();
    auto second = b->receive();
    require(first.kind == Kind::text && first.data == "first", "first frame changed");
    require(second.kind == Kind::binary && second.data == std::string("\0\xff", 2), "binary frame changed");
    require(a->receive().data == "reverse", "reverse direction did not arrive");
    a->send({Kind::text, "before close"});
    a->close(4012, "finished");
    require(b->receive().data == "before close", "close discarded an earlier frame");
    try { b->receive(); throw std::runtime_error("close was not delivered"); }
    catch (const CloseError& error) {
        require(error.code == 4012 && error.reason == "finished", "close changed code or reason");
    }
    refuses<Closed>([&] { a->send({Kind::text, "late"}); });
    refuses<Closed>([&] { a->receive(); });
}

void pacing() {
    auto [a, b] = pipe(1024, 1);
    a->send({Kind::text, "one"});
    auto sending = std::async(std::launch::async, [&] { a->send({Kind::text, "two"}, Wait::after(2s)); });
    require(sending.wait_for(25ms) == std::future_status::timeout, "full transport did not pace the sender");
    require(b->receive().data == "one", "queued frame changed");
    sending.get();
    require(b->receive().data == "two", "paced frame did not arrive");
    refuses<DeadlineExceeded>([&] { b->receive(Wait::after(10ms)); });
    a->send({Kind::text, "still usable"});
    require(b->receive().data == "still usable", "a receive timeout killed the connection");
}

void cancellation_and_abort() {
    auto [a, b] = pipe(1024, 1);
    std::stop_source source;
    auto receiving = std::async(std::launch::async, [&] {
        refuses<Cancelled>([&] { b->receive(Wait{source.get_token()}); });
    });
    source.request_stop();
    receiving.get();
    a->send({Kind::text, "one"});
    auto sending = std::async(std::launch::async, [&] {
        refuses<Closed>([&] { a->send({Kind::text, "blocked"}); });
    });
    a->abort();
    require(sending.wait_for(1s) == std::future_status::ready, "abort left send blocked");
    sending.get();
    require(b->receive().data == "one", "abort discarded a previously sent frame");
    try { b->receive(); throw std::runtime_error("abort was not delivered"); }
    catch (const CloseError& error) { require(error.code == 1006, "abort was not abnormal closure"); }
}

void byte_limit() {
    auto [a, b] = pipe(3);
    a->send({Kind::text, "\xf0\x9f\x98\x80"});
    refuses<FrameTooLarge>([&] { b->receive(); });
    refuses<Closed>([&] { b->receive(); });
    refuses<CloseError>([&] { a->send({Kind::text, "late"}); });
}

int main() {
    try { seam(); pacing(); cancellation_and_abort(); byte_limit(); }
    catch (const std::exception& error) { std::cerr << error.what() << '\n'; return 1; }
}

File: duplex/rs/tests/seam.rs
use nightseam_duplex::{Frame, pipe};
use std::time::Duration;
use tokio::time::timeout;

#[tokio::test]
async fn frames_are_whole_ordered_and_bidirectional() {
    let (a, b) = pipe(1024);
    for frame in [
        Frame::Text(vec![]),
        Frame::Text("hé😀".as_bytes().to_vec()),
        Frame::Binary((0..=255).collect()),
    ] {
        a.send(frame.clone()).await.unwrap();
        assert_eq!(b.receive().await.unwrap(), frame);
        b.send(frame.clone()).await.unwrap();
        assert_eq!(a.receive().await.unwrap(), frame);
    }
    a.send(Frame::Text(b"first".to_vec())).await.unwrap();
    a.send(Frame::Text(b"second".to_vec())).await.unwrap();
    a.close(4001, "finished").await.unwrap();
    assert_eq!(b.receive().await.unwrap(), Frame::Text(b"first".to_vec()));
    assert_eq!(b.receive().await.unwrap(), Frame::Text(b"second".to_vec()));
    let ended = b.receive().await.unwrap_err();
    assert_eq!((ended.code, ended.reason.as_str()), (4001, "finished"));
    assert!(a.send(Frame::Binary(vec![])).await.is_err());
    assert!(b.send(Frame::Binary(vec![])).await.is_err());
}

#[tokio::test]
async fn receive_limit_prevents_delivery_and_ends_the_pipe() {
    let (a, b) = pipe(2);
    a.send(Frame::Text(b"big".to_vec())).await.unwrap();
    assert!(b.receive().await.is_err());
    assert_eq!(a.receive().await.unwrap_err().code, 1006);
    assert!(b.send(Frame::Binary(vec![])).await.is_err());
}

#[tokio::test]
async fn pipe_paces_after_eight_frames_and_abort_releases_both_directions() {
    let (a, b) = pipe(1024);
    for _ in 0..8 {
        a.send(Frame::Binary(vec![1])).await.unwrap();
    }
    assert!(
        timeout(Duration::from_millis(20), a.send(Frame::Binary(vec![2])))
            .await
            .is_err()
    );
    assert_eq!(b.receive().await.unwrap(), Frame::Binary(vec![1]));
    a.send(Frame::Binary(vec![3])).await.unwrap();
    let blocked = {
        let a = a.clone();
        tokio::spawn(async move { a.send(Frame::Binary(vec![4])).await })
    };
    a.abort();
    assert!(
        timeout(Duration::from_secs(1), blocked)
            .await
            .unwrap()
            .unwrap()
            .is_err()
    );
    for _ in 0..8 {
        assert!(b.receive().await.is_ok());
    }
    assert_eq!(b.receive().await.unwrap_err().code, 1006);
    assert!(a.receive().await.is_err());
}
